### DISCUSSION TITLE
Auto scaling on k8s env

### DISCUSSION THREAD

**Discussion by janiu-001** [emqx/emqx#8889] [Q&A]
Hi Expert, 

Just wanna confirm how could we support auto scaling in k8s cluster? Very appreciated for your response

And Could use the free version for commercial usage?

Br
Baoke

**Comment by zmstone** [comment]
Hi @janiu-001 

The main challenge for autoscaling is: EMQX cluster is stateful in a way that all nodes needs to have the full view of the cluster.

If you run EMQX 5.0, the cluster can be configured with two different roles, core and replicant.
The 'replicant' nodes are quite stateless to fit in k8s auto-scaling, 
they do no need to have to full view of the cluster hence can be added and deleted more freely.
You can find more details here: https://www.emqx.com/en/blog/how-emqx-5-0-achieves-100-million-mqtt-connections

If you run EMQX 4.x, new nodes added when scaling up will need to 'join' the cluster.
and old nodes to be removed before scaling down will need to 'leave' the cluster,
(or removed from the cluster by one of the peer nodes applying a 'foce_leave').
